mathics/eval/tensors.py

```diff
@@ -76,6 +74,88 @@ def get_dimensions(expr, head=None):
         return [len(expr.elements)] + sub
 
 
+def to_std_sparse_array(sparse_array, evaluation: Evaluation):
+    if sparse_array.elements[2] == Integer0:
+        return sparse_array
+    else:
+        return Expression(
+            SymbolSparseArray, Expression(SymbolNormal, sparse_array)
+        ).evaluate(evaluation)
+
+
+def unpack_outer(item, rest_lists, current, level: int, const_etc: tuple):
+    """
+    Recursively unpacks lists to evaluate outer product.
+    ------------------------------------
+
+    Unlike direct products, outer (tensor) products require traversing the
+    lowest level of each list, hence we recursively unpacking lists until
+    the lowest level is reached.
+
+    Parameters:
+
+    ``item``: the current item to be unpacked (if not at lowest level), or joined
+    to current (if at lowest level)
+
+    ``rest_lists``: the rest of lists to be unpacked
+
+    ``current``: the current lowest level elements
+
+    ``level``: the current level (unused yet, will be used in
+    ``Outer[f_, lists__, n_]`` in the future)
+
+    ``const_etc``: a tuple of functions used in unpacking, remains constant
+    throughout the recursion.
+
+    Format of ``const_etc``:
+
+    ```
+    (
+        cond_next_list,  # return True/False to unpack the next list/this list at next level
+        get_elements,  # get elements of list, tuple, ListExpression, etc.
+        apply_head,  # e.g. lambda elements: Expression(head, *elements)
+        apply_f,  # e.g. lambda current: Expression(f, *current)
+        join_elem,  # join current lowest level elements (i.e. current) with a new one
+        if_nested,  # True for result as nested list, False for result as flattened list
+        evaluation,  # evaluation: Evaluation
+    )
+    ```
+    """
+    (

```

Comment:
@Li-Xiang-Ideal, is there any advantage in using a magic tuple instead of a dictionary to pass these parameters?

Reply:
Actually, the better thing to use here is what is called [NamedTuple](https://docs.python.org/3.11/library/typing.html?highlight=namedtuple#typing.NamedTuple)

Like a tuple this is immutable, which is what we want for clarity. And like a dictionary access is via a meaningful name. However this is accessed as an attribute of the object rather than as a key, which is also considered by those who are into the style kind of thing.

Reply:
@rocky, a NamedTuple would make sense to me if the bunch of functions were built in some place and then passed to other functions. Here it is just a bunch of parameters packed to be passed to a function, whose first instruction is to unpack it in local variables. For some reason, the standard structure used to do this in Python is the `dict`. Am I missing something?

Reply:
@Li-Xiang-Ideal, this was more a question for @rocky. Again, for me it is ok to merge this as it is now.

Reply:
> Here it is just a bunch of parameters packed to be passed to a function, whose first instruction is to unpack it in local variables

IMHO then a tuple is fine and it happens to be efficient too.  

Use `*args`,  and `**kwargs` is frowned upon more nowadays except in very rare and custom circumstances.

One reason for this is that nowadays type annotations are desirable and used more by linters if not other kinds of things. 

I would not be surprised if, in the future, instead of being ignored in running the code, that they are used to speed up code as you find in mypy. This seems the direction that 3.12 is going towards. Because Guido is no longer in the center of things as before, attitudes have been shifting (in my opinion for the better). 

The problem with `*args`, and `**kwargs` is that these can't be type checked.

Reply:
> Here apply_head is identity function, so return apply_head(elements) will give a list instead of our M-Expression like ListExpression. This is because rules in SparseArray is a flattened list of pos -> val. So I set if_nested=False to automatically get a flattened list with elements.extend() and apply ListExpression manually in the end. 

Thanks for this clarification. Putting this is as a comment in the code would be helpful.